uaclient/daemon/retry_auto_attach.py

```
@@ -0,0 +1,228 @@
+    elif isinstance(e, api_exceptions.NonAutoAttachImageError):
+        return "Canonical servers did not recognize this image as Ubuntu Pro"
+    elif isinstance(e, api_exceptions.LockHeldError):
+        return "the pro lock was held by pid {pid}".format(pid=e.pid)
+    elif isinstance(e, api_exceptions.ContractAPIError):
+        return 'an error from Canonical servers: "{}"'.format(e.api_error)
+    elif isinstance(e, api_exceptions.UrlError):
+        if e.url:
+            if e.code:
+                failure_reason = "a {code} while reaching {url}".format(
+                    code=e.code, url=e.url
+                )
+            else:
+                failure_reason = "an error while reaching {url}".format(
+                    url=e.url
+                )
+        else:
+            failure_reason = "a network error"
+        failure_reason += ': "{}"'.format(str(e))
+        return failure_reason
+    elif isinstance(e, api_exceptions.UserFacingError):
+        return '"{}"'.format(e.msg)
+    else:
+        LOG.error("Unexpected exception: {}".format(e))
+        return str(e) or "an unknown error"
+
+
+def _try_full_auto_attach():
+    persisted_options = OPTIONS_FILE.read()
+    for i in range(2):
+        options = FullAutoAttachOptions()
+        if persisted_options is not None:
+            options.enable = persisted_options.enable
+            options.enable_beta = persisted_options.enable_beta
+        try:
+            full_auto_attach(options)
+            return
+        except (
+            api_exceptions.BetaServiceError,
+            api_exceptions.EntitlementNotFoundError,
+            api_exceptions.IncompatibleEntitlementsDetected,
+        ) as e:
+            LOG.warning(
+                "auto attach failed because of args. error message: {}".format(
+                    e.msg
+                )
+            )
+            if persisted_options is not None:
+                LOG.info("trying again with defaults")
```

Sorry, I may have missed or forgotten about a discussion around this, but isn't it risk if we `auto-attach` without the user options ? We could enable unwanted services on this setup right ?

---

Let's meet and discuss

---

We decided that the full_auto_attach api function should actually do a best-effort attempt to attach and enable services instead of failing early if it detects a bad argument. 
That change will make this fallback to auto-attach without user options unnecessary